crates/db_connection_pool/src/dbconnection/snowflakeconn.rs

```diff
@@ -145,3 +169,144 @@ impl<'a> AsyncDbConnection<Arc<SnowflakeApi>, &'a (dyn Sync)> for SnowflakeConne
     DataFusionError::Execution(format!("{}", e.into()).to_string())
 }
+
+/// Converts `Snowflake` specific types to standard Arrow types.
+///
+/// # Errors
+///
+/// Returns an error if there is a failure in converting Snowflake to Arrow types.
+pub fn snowflake_schema_cast(record_batch: &RecordBatch) -> Result<RecordBatch, Error> {
+    let mut fields = Vec::new();
+    let mut columns = Vec::new();
+
+    for (idx, field) in record_batch.schema().fields().iter().enumerate() {
+        let column = record_batch.column(idx);
+        if let Some(sf_logical_type) = field.metadata().get("logicalType") {
+            if sf_logical_type.to_lowercase().as_str() == "timestamp_ntz" {
+                fields.push(Arc::new(Field::new(
+                    field.name(),
+                    DataType::Timestamp(TimeUnit::Millisecond, None),
+                    field.is_nullable(),
+                )));
+                columns.push(cast_sf_timestamp_ntz_to_arrow_timestamp(column)?);
+                continue;
+            }
+        }
+        fields.push(Arc::clone(field));
+        columns.push(Arc::clone(column));
+    }
+
+    let schema = Arc::new(Schema::new(fields));
+    RecordBatch::try_new(schema, columns).context(FailedToCreateRecordBatchSnafu)
+}
+
+fn cast_sf_timestamp_ntz_to_arrow_timestamp(column: &ArrayRef) -> Result<ArrayRef, Error> {
+    let struct_array = column.as_any().downcast_ref::<StructArray>().context(
+        UnableToCastSnowflakeTimestampSnafu {
+            reason: "value is not a struct",
+        },
+    )?;
+    let epoch_array = struct_array
+        .column(0)
+        .as_any()
+        .downcast_ref::<Int64Array>()
+        .context(UnableToCastSnowflakeTimestampSnafu {
+            reason: "epoch is missing",
+        })?;
+    let fraction_array = struct_array
+        .column(1)
+        .as_any()
+        .downcast_ref::<Int32Array>()
+        .context(UnableToCastSnowflakeTimestampSnafu {
+            reason: "fraction is missing",
+        })?;
+
+    let mut builder = TimestampMillisecondBuilder::new();
+
+    for idx in 0..struct_array.len() {
+        if struct_array.is_null(idx) {
+            builder.append_null();
+        } else {
+            let epoch = epoch_array.value(idx);
+            let fraction = i64::from(fraction_array.value(idx));
+            let timestamp = epoch * 1_000 + fraction / 1_000_000;
+            builder.append_value(timestamp);
+        }
+    }
+    Ok(Arc::new(builder.finish()) as ArrayRef)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use arrow::array::{
+        ArrayBuilder, ArrayRef, Int32Builder, Int64Builder, StructBuilder,
+        TimestampMillisecondArray,
+    };
+    use arrow::datatypes::{DataType, Field};
+    use std::sync::Arc;
+
+    #[test]
+    fn test_cast_sf_timestamp_ntz_to_arrow_timestamp() {
+        let timestamp_ntz_array = create_timestamp_ntz_array(
+            vec![Some(1_696_164_330), None, Some(1_714_647_301)],
```

Comment:
I'd test all of the permutations here.  (Some, Some)(none, none), (Some, None), (None, Some)

Reply:
Added test for invalid input 👍
https://github.com/spiceai/spiceai/pull/1677/files#diff-e0b90357e2dacc69cb483074b1b4a82a0d58da4d111f3f0179589a5053071b41R276